### PR TITLE
Avoid double counting in `stats` `out` when writing to nested streams

### PR DESCRIPTION
**Pull request by nhz2**
Here is an example of the bug this PR fixes:

```julia
using Test
using TranscodingStreams
stream = NoopStream(NoopStream(IOBuffer()))
write(stream, b"foobar")
flush(stream)
@test TranscodingStreams.stats(stream).out == 6
```
Before this PR this would output 12 even though only 6 bytes were written to the underlying stream.

### CODE REVIEW

File: src/stream.jl [hunk 1]
```diff
@@ -489,21 +489,21 @@ function Base.write(stream::TranscodingStream)
     return 0
 end
 
-function Base.write(stream::TranscodingStream, b::UInt8)
+function Base.write(stream::TranscodingStream, b::UInt8)::Int
     changemode!(stream, :write)
-    if marginsize(stream.buffer1) == 0 && flushbuffer(stream) == 0
-        return 0
```
Comment (nhz2) [inline src/stream.jl:495]: `write` will throw an error or block instead of returning 0 on failure.

File: src/stream.jl [hunk 2]
```diff
@@ -489,21 +489,21 @@ function Base.write(stream::TranscodingStream)
...
 
 function Base.unsafe_write(stream::TranscodingStream, input::Ptr{UInt8}, nbytes::UInt)
     changemode!(stream, :write)
-    state = stream.state
     buffer1 = stream.buffer1
     p = input
     p_end = p + nbytes
-    while p < p_end && (marginsize(buffer1) > 0 || flushbuffer(stream) > 0)
+    while p < p_end
+        marginsize(buffer1) > 0 || flush_buffer1(stream)
```
Comment (nhz2) [inline src/stream.jl:506]: `flush_buffer1` must empty `buffer1` or error, so this loop should make progress.

File: src/stream.jl [hunk 3]
```diff
@@ -595,9 +595,12 @@ function stats(stream::TranscodingStream)
         out = transcoded_out - buffersize(buffer1)
     elseif mode === :write
         transcoded_in = buffer1.transcoded
-        transcoded_out = buffer2.transcoded
```
Comment (nhz2) [inline src/stream.jl:598]: `buffer2` is "owned" by the underlying stream if it is shared, so this stream shouldn't rely on `buffer2.transcoding` being the expected value.

**Comment by nhz2** [comment]
@reallyasi9 Thank you for updating ZipStreams.jl for this fix. I'll merge this PR sometime next week unless there are other concerns.

**Review by mkitti — APPROVED** [review]

**Pull request merged**